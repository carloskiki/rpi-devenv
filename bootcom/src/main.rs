use std::{
    io::{copy, stdout, Write},
    time::Duration,
};

use serialport::{DataBits, FlowControl, Parity, StopBits};

fn main() {
    let mut args = std::env::args();
    args.next(); // Skip the command name
    let port_name = args.next().expect("a port name should be provided");
    let binary_name = args.next().expect("a binary file should be provided");

    println!("=> Connecting to the serial port `{port_name}` ...");
    println!("=> Each operation on the port times out after 60 seconds.");
    let mut port = serialport::new(port_name, 115200)
        .data_bits(DataBits::Eight)
        .parity(Parity::None)
        .flow_control(FlowControl::None)
        .stop_bits(StopBits::One)
        .timeout(Duration::from_secs(60))
        .open()
        .expect("the serial port provided should be valid");
    println!("=> Connected to the serial port.");

    println!("=> Waiting for the device to be ready ...");
    let mut ready_buf = [0u8; 1];
    port.read_exact(&mut ready_buf)
        .expect("should be able to read from the serial port");
    assert_eq!(
        ready_buf[0], 0xff,
        "The device needs to output 0xff to indicate it is ready"
    );
    println!("=> The device is ready.");

    println!("=> Sending the binary ...");
    let binary = std::fs::read(binary_name).expect("the binary file should be readable");
    let binary_size: u32 = binary
        .len()
        .try_into()
        .expect("the binary file should be less than 2^32 bytes");
    port.write_all(&binary_size.to_le_bytes())
        .expect("should be able to write to the serial port");
    port.write_all(&binary)
        .expect("should be able to write to the serial port");
    println!("=> The binary has been sent.");

    port.set_timeout(Duration::from_secs(5))
        .expect("should be able to set the timeout");
    let _ = copy(&mut port, &mut stdout());
    stdout().flush().expect("should be able to flush stdout");
    println!("=> The device has finished executing.");
}
