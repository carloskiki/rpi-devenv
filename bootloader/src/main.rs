#![no_std]
#![no_main]
use rpi::gpio::{Alternate5, Pin};

core::arch::global_asm!(include_str!("boot.s"), options(raw));

#[no_mangle]
pub extern "C" fn first_stage() -> ! {
    let mut uart = rpi::uart::MiniUart::get().unwrap();
    uart.set_bit_mode(true);
    uart.set_baud_rate(115200);
    let tx_pin: Pin<14, Alternate5> = Pin::get().unwrap();
    let rx_pin: Pin<15, Alternate5> = Pin::get().unwrap();
    let mut rx_tx = uart.enable_transmitter(tx_pin).enable_receiver(rx_pin);
    rx_tx.send_blocking([0xff].into_iter());

    let mut binary_size_bytes = [0u8; 4];
    rx_tx.receive_exact(&mut binary_size_bytes);
    let mut binary_size = u32::from_le_bytes(binary_size_bytes);

    let mut byte_buf = [0u8];
    // Safety: This value is provided by the linker script.
    unsafe extern "C" {
        #[link_name = "__physical_load_address"]
        static LOAD_ADDRESS: u8;
    }
    // Safety: We know that there is nothing at __physical_load_address, since we relocated to
    // 0x2000000 (__relocate_address).
    unsafe {
        let mut dest_ptr: *mut u8 = (&raw const LOAD_ADDRESS) as _;
        while binary_size > 0 {
            rx_tx.receive_exact(&mut byte_buf);

            dest_ptr.write(byte_buf[0]);
            dest_ptr = dest_ptr.offset(1);
            binary_size -= 1;
        }
    }

    // Clean up before jumping to the kernel
    drop(rx_tx);

    // Safety: We know that the kernel is a function that never returns.
    // We also have loaded it into memory at LOAD_ADDRESS.
    unsafe {
        let kernel: fn() -> ! = core::mem::transmute(&raw const LOAD_ADDRESS);
        kernel();
    }
}

#[panic_handler]
fn panic(_info: &core::panic::PanicInfo) -> ! {
    loop {}
}
