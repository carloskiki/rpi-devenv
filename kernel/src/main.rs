#![no_std]
#![no_main]

use core::arch::asm;
use core::{arch::global_asm, hint::black_box};

use rpi::mmu::TRANSLATION_TABLE;
use rpi::{data_memory_barrier, data_synchronization_barrier};

// const GPFSEL3: usize = 0x2020000C;
const GPFSEL4: usize = 0x20200010;
const GPSET1: usize = 0x20200020;
const GPCLR1: usize = 0x2020002C;

const BOOT_ADDRESS: usize = 0x8000;

global_asm!(include_str!("boot.s"),
    ABORT_MODE = const ProcessorMode::Abort as u32,
    SVC_MODE = const ProcessorMode::Supervisor as u32
);

/// Things we do in the first stage:
/// - Set up and enable the MMU
/// - Change the processor mode to system mode
#[unsafe(no_mangle)]
pub extern "C" fn first_stage() -> ! {
    // let mut uart = MiniUart::get().unwrap();
    // uart.set_bit_mode(true);
    // uart.set_baud_rate(115200);
    // let tx_pin: Pin<14, Alternate5> = Pin::get().unwrap();
    // let rx_pin: Pin<15, Alternate5> = Pin::get().unwrap();
    // let mut rx_tx = uart.enable_transmitter(tx_pin).enable_receiver(rx_pin);
    // rx_tx.send_blocking("hello world\n".bytes());

    unsafe {
        let addr = (TRANSLATION_TABLE.0.as_ptr() as usize) & (!0 << 14) | 0b11011;

        // Set the translation table base address with its flags
        asm!("mcr p15, 0, {}, c2, c0, 0", in(reg) addr, options(nostack, nomem, preserves_flags));
        // Set the value of `N` to 0 in the translation table base control register
        asm!("mcr p15, 0, {}, c2, c0, 2", in(reg) 0, options(nostack, nomem, preserves_flags));
        // Set all domains to manager mode
        asm!("mcr p15, 0, {}, c3, c0, 0", in(reg) !0, options(nostack, nomem, preserves_flags));

        // Invalidate caches and TLB
        asm!("mcr p15, 0, {0}, c7, c7, 0
              mcr p15, 0, {0}, c8, c7, 0", in(reg) 0, options(nostack, nomem, preserves_flags));
        // bit 12: enable L1 instruction cache
        // bit 11: enable branch prediction
        // bit 2: enable L1 data cache
        // bit 0: enable MMU
        let mut control_reg: u32;
        asm!("mrc p15, 0, {}, c1, c0, 0", out(reg) control_reg, options(nostack, nomem, preserves_flags));
        control_reg |= 0b101;
        control_reg |= 0b11 << 11;
        asm!("mcr p15, 0, {}, c1, c0, 0", in(reg) control_reg, options(nostack, nomem, preserves_flags));
    }
    data_synchronization_barrier();

    loop {}
}

#[repr(u8)]
enum ProcessorMode {
    User = 0b10000,
    Fiq = 0b10001,
    Irq = 0b10010,
    Supervisor = 0b10011,
    Abort = 0b10111,
    Undefined = 0b11011,
    System = 0b11111,
}

#[panic_handler]
fn panic(_info: &core::panic::PanicInfo) -> ! {
    unsafe fn get32(addr: usize) -> u32 {
        core::ptr::read_volatile(addr as *const u32)
    }

    unsafe fn put32(adder: usize, value: u32) {
        core::ptr::write_volatile(adder as *mut u32, value);
    }
    
    const BLINK_DELAY: u32 = 0x400000;
    fn delay(mut n: u32) {
        while n > 0 {
            n -= 1;
            black_box(n);
        }
    }

    data_memory_barrier();

    // Set GPIO pin 47 as output
    unsafe {
        get32(GPFSEL4);
        let mask = 0b111 << 21;
        let output = 0b001 << 21;
        put32(GPFSEL4, (get32(GPFSEL4) & !mask) | output);
    }

    loop {
        // Turn off the LED
        unsafe {
            put32(GPSET1, 1 << 15);
        }
        delay(BLINK_DELAY);

        // Turn on the LED
        unsafe {
            put32(GPCLR1, 1 << 15);
        }
        delay(BLINK_DELAY);
    }
}

// IMPORTANT: You only have 16KiB of stack space. Do not use more than that.
#[unsafe(no_mangle)]
pub unsafe extern "C" fn interrupt_handler() {
    // TODO: ...
}

/// The interrupt handler for Data Abort, Prefecth Abort, and Undefined Instructions.
#[unsafe(no_mangle)]
pub unsafe extern "C" fn interrupt_panic() -> ! {
    panic!()
}
