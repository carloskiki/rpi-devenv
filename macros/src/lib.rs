use proc_macro::TokenStream as TokenStream1;
use proc_macro2::TokenStream;
use quote::quote;
use syn::{parse_macro_input, Error, ItemFn, ReturnType, Signature, Type};

/// The `#[main]` attribute macro, which is used to mark the entry point of the program.
///
/// This macro must be attached to a function that takes no arguments, and never returns: () -> !.
#[proc_macro_attribute]
pub fn main(args: TokenStream1, input: TokenStream1) -> TokenStream1 {
    if !args.is_empty() {
        return Error::new_spanned(TokenStream::from(args), "this macro takes no arguments")
            .into_compile_error()
            .into();
    }

    match modify_fn(parse_macro_input!(input as ItemFn)) {
        Ok(stream) => stream.into(),
        Err(err) => err.into_compile_error().into(),
    }
}

fn modify_fn(
    ItemFn {
        attrs,
        vis,
        sig:
            Signature {
                constness,
                asyncness,
                unsafety,
                abi,
                fn_token,
                ident,
                generics,
                inputs,
                variadic,
                output,
                ..
            },
        block,
    }: ItemFn,
) -> Result<TokenStream, Error> {
    if asyncness.is_some() {
        return Err(Error::new_spanned(
            asyncness,
            "main function cannot be async",
        ));
    }
    if unsafety.is_some() {
        return Err(Error::new_spanned(
            unsafety,
            "main function cannot be unsafe",
        ));
    }
    if abi.is_some() {
        return Err(Error::new_spanned(
            abi,
            "main function must use the default ABI",
        ));
    }
    if !generics.params.is_empty() {
        return Err(Error::new_spanned(
            generics,
            "main function cannot have generic parameters",
        ));
    }
    if !inputs.is_empty() {
        return Err(Error::new_spanned(
            inputs,
            "main function cannot have parameters",
        ));
    }
    if variadic.is_some() {
        return Err(Error::new_spanned(
            variadic,
            "main function cannot be variadic",
        ));
    }
    let ReturnType::Type(rt_token, return_type) = output else {
        return Err(Error::new_spanned(
            output,
            "main function must never return (`-> !`)",
        ));
    };
    let &Type::Never(_) = return_type.as_ref() else {
        return Err(Error::new_spanned(
            return_type,
            "main function return type must be `!`",
        ));
    };
    let output = ReturnType::Type(rt_token, return_type);
    Ok(quote! {
        #(#attrs)*
        #[unsafe(export_name = "_main")]
        #vis #constness #fn_token #ident () #output {
            #block
        }
    })
}
