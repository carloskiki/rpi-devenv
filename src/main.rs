#![no_std]
#![no_main]

use core::{arch::global_asm, hint::black_box};

use rpi::mem_barrier;
use rpi::uart::MiniUart;
use rpi::gpio::{Alternate5, Pin};

// const GPFSEL3: usize = 0x2020000C;
const GPFSEL4: usize = 0x20200010;
const GPSET1: usize = 0x20200020;
const GPCLR1: usize = 0x2020002C;

const TIMER_FOUR_SEC: u32 = 0x400000;

global_asm!(include_str!("boot.s"), options(raw));

#[no_mangle]
pub extern "C" fn first_stage() -> ! {
    // Set GPIO pin 47 as output
    unsafe {
        get32(GPFSEL4);
        let mask = 0b111 << 21;
        let output = 0b001 << 21;
        put32(GPFSEL4, (get32(GPFSEL4) & !mask) | output);
    }

    // Turn on the LED
    unsafe {
        put32(GPCLR1, 1 << 15);
    }

    // Safety: QEMU being a bitch
    let mut uart = unsafe { MiniUart::get_unlocked() };
    uart.set_bit_mode(true);
    uart.set_baud_rate(115200);
    let tx_pin: Pin<14, Alternate5> = Pin::get().unwrap();
    let rx_pin: Pin<15, Alternate5> = Pin::get().unwrap();
    let mut rx_tx = uart
        .enable_transmitter(tx_pin)
        .enable_receiver(rx_pin);
    rx_tx.send_blocking("hello world\n".bytes());
    let mut byte = [0];
    loop {
        rx_tx.receive_exact(&mut byte[..]);
        rx_tx.send_blocking(byte.iter().copied());
    }
}

#[repr(u8)]
enum ProcessorMode {
    User = 0b10000,
    Fiq = 0b10001,
    Irq = 0b10010,
    Supervisor = 0b10011,
    Abort = 0b10111,
    Undefined = 0b11011,
    System = 0b11111,
}

unsafe fn get32(addr: usize) -> u32 {
    core::ptr::read_volatile(addr as *const u32)
}

unsafe fn put32(adder: usize, value: u32) {
    core::ptr::write_volatile(adder as *mut u32, value);
}

fn delay(mut n: u32) {
    while n > 0 {
        n -= 1;
        black_box(n);
    }
}

#[export_name = "rust_irq_handler"]
pub extern "C" fn irq_handler() {}

#[panic_handler]
fn panic(_info: &core::panic::PanicInfo) -> ! {
    unsafe { mem_barrier() };

    loop {
        // Turn off the LED
        unsafe {
            put32(GPSET1, 1 << 15);
        }
        delay(TIMER_FOUR_SEC);

        // Turn on the LED
        unsafe {
            put32(GPCLR1, 1 << 15);
        }
        delay(TIMER_FOUR_SEC);
    }
}

