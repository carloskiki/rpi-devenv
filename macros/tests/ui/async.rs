use macros::main;

#[main]
async fn kernel_main() -> ! {
    todo!()
}

fn main() {}
