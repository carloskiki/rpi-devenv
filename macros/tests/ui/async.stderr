error: main function cannot be async
 --> tests/ui/async.rs:4:1
  |
4 | async fn kernel_main() -> ! {
  | ^^^^^
